### PR TITLE
add paddle.nn.SmoothL1Loss, test=develop

### PR DESCRIPTION
**Pull request by yghstill**
### PR types
 New features

### PR changes
 APIs 

### Describe
add paddle.nn.SmoothL1Loss. This operator is calculate smooth_l1_loss. Creates a criterion that uses a squared term if the absolute element-wise error falls below 1 and an L1 term otherwise. In some cases it can prevent exploding gradients. Also known as the Huber loss.

class:
```python
class paddle.nn.SmoothL1Loss(reduction='mean')
```
functioanl:
```python
def paddle.nn.functioanl.smooth_l1_loss(input, label)
```
docs
image
image
image

Examples:
```python
           # declarative mode
            import paddle
            import paddle.fluid as fluid
            import numpy as np
            input = fluid.layers.data(name="input", shape=[-1, 3], dtype="float32")
            label = fluid.layers.data(name="label", shape=[-1, 3], dtype="float32")
            loss = paddle.nn.SmoothL1Loss()
            result = loss(x,label)
            place = fluid.CPUPlace()
            exe = fluid.Executor(place)
            exe.run(fluid.default_startup_program())
            input_data = np.random.rand(3,3).astype("float32")
            label_data = np.random.rand(3,3).astype("float32")
            output= exe.run(feed={"input": input_data, "label": label_data},
                            fetch_list=[result])
            print(output)

            # imperative mode
            import paddle.fluid.dygraph as dg
            with dg.guard(place) as g:
                input = dg.to_variable(input_data)
                label = dg.to_variable(label_data)
                weight = dg.to_variable(weight_data)
                loss = paddle.nn.SmoothL1Loss()
                output = loss(input, label)
                print(output.numpy())
```

### CODE REVIEW

File: paddle/fluid/operators/smooth_l1_loss_v2_op.cc [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+             "This input is optional and should have same shape with X. "
+             "If provided, the out smooth l1 loss will be multiplied by this "
+             "tensor element by element.")
+        .AsDispensable();
+    AddOutput("Diff", "Intermediate variable to cache InsideWeight * (X - Y).")
+        .AsIntermediate();
+    AddOutput("Out",
+              "(Tensor, default Tensor<float>) A tensor with rank be 2. "
+              "The output smooth l1 loss has same dimension as Input(X).");
+    AddAttr<float>("sigma",
```
Comment (jerrywgz) [inline paddle/fluid/operators/smooth_l1_loss_v2_op.cc:117]: default value is different with comments

File: paddle/fluid/operators/smooth_l1_loss_v2_op.h [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    auto* out0 = context.Output<Tensor>("Diff");
+    auto* out1 = context.Output<Tensor>("Out");
+
+    out0->mutable_data<T>(context.GetPlace());
+    out1->mutable_data<T>(context.GetPlace());
+    auto* place =
+        context.template device_context<DeviceContext>().eigen_device();
+
+    auto sigma = static_cast<T>(context.Attr<AttrType>("sigma"));
+    T sigma2 = sigma * sigma;
```
Comment (jerrywgz) [inline paddle/fluid/operators/smooth_l1_loss_v2_op.h:64]: need to check whether sigma is same as torch

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -5830,6 +5831,82 @@ def smooth_l1(x, y, inside_weight=None, outside_weight=None, sigma=None):
...
+	:old_api: paddle.fluid.layers.smooth_l1_v2
+
+    This layer computes the smooth L1 loss for Variable :attr:`x` and :attr:`y`.
+    It takes the first dimension of :attr:`x` and :attr:`y` as batch size.
+    For each instance, it computes the smooth L1 loss element by element first
+    and then sums all the losses. So the shape of output Variable is
+    [batch_size, 1].
+
+    Args:
+        x (Variable): A tensor with rank at least 2. The input value of smooth
```
Comment (jerrywgz) [inline python/paddle/fluid/layers/nn.py]: Modify Variable to Tensor

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -5830,6 +5831,82 @@ def smooth_l1(x, y, inside_weight=None, outside_weight=None, sigma=None):
...
+            input is optional and should have same shape with :attr:`x`. If
+            provided, the result of (:attr:`x` - :attr:`y`) will be multiplied
+            by this tensor element by element.
+            A Tensor with type float32.
+        outside_weight (Variable|None): A tensor with rank at least 2. This
+            input is optional and should have same shape with :attr:`x`. If
+            provided, the out smooth L1 loss will be multiplied by this tensor
+            element by element.
+            A Tensor with type float32.
+        sigma (float|None): Hyper parameter of smooth L1 loss layer. A float
```
Comment (jerrywgz) [inline python/paddle/fluid/layers/nn.py]: default value is 1.0 or None?

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -5830,6 +5831,82 @@ def smooth_l1(x, y, inside_weight=None, outside_weight=None, sigma=None):
...
+           scalar with default value 1.0.
+
+    Returns:
+        Variable: The output smooth L1 loss shape is the same as the shape of x(input).  A Tensor with type float32.
```
Comment (jerrywgz) [inline python/paddle/fluid/layers/nn.py]: Variable -> Tensor

File: python/paddle/fluid/tests/unittests/test_smooth_l1_loss.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (jerrywgz) [inline python/paddle/fluid/tests/unittests/test_smooth_l1_loss.py:1]: no need for this unittest?

File: paddle/fluid/operators/smooth_l1_loss_v2_op.cc (hunk 1 above)
Comment (yghstill) [reply paddle/fluid/operators/smooth_l1_loss_v2_op.cc:117]: done,thx.

File: paddle/fluid/operators/smooth_l1_loss_v2_op.h (hunk 2 above)
Comment (yghstill) [reply paddle/fluid/operators/smooth_l1_loss_v2_op.h:64]: not use sigma in torch, sigma is to adapt the history of paddle.

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (yghstill) [reply python/paddle/fluid/layers/nn.py]: done, thx

File: python/paddle/fluid/tests/unittests/test_smooth_l1_loss.py (hunk 6 above)
Comment (yghstill) [reply python/paddle/fluid/tests/unittests/test_smooth_l1_loss.py:1]: smooth_l1_loss class need unittest

**Review by jerrywgz — DISMISSED** [review]
LGTM

**Review by jerrywgz — DISMISSED** [review]
LGTM

**Review by jerrywgz — DISMISSED** [review]
LGTM

**Review by yaoxuefeng6 — COMMENTED** [review]
fluid 目录下不必新增V2版本api 直接在paddle目录下新增一个api即可

**Comment by yghstill** [comment]
> fluid 目录下不必新增V2版本api 直接在paddle目录下新增一个api即可

@yaoxuefeng6 Done，thx.

**Review by yaoxuefeng6 — DISMISSED** [review]
LGTM

**Review by jerrywgz — DISMISSED** [review]
LGTM

**Review by Shixiaowei02 — DISMISSED** [review]

**Review by juncaipeng — DISMISSED** [review]
LGTM

File: python/paddle/nn/functional/loss.py [hunk 7]
```diff
@@ -64,13 +64,116 @@
...
     'square_error_cost',
     'ssd_loss',
     'teacher_student_sigmoid_loss'
 ]
 
 
+def smooth_l1_loss(input, label, reduction='mean'):
+    """
+	:alias_main: paddle.nn.functional.smooth_l1_loss
+	:alias: paddle.nn.functional.smooth_l1_loss,paddle.nn.functional.loss.smooth_l1_loss
```
Comment (saxon-zh) [inline python/paddle/nn/functional/loss.py]: alias标签不需要自己来写了，可以删除

File: python/paddle/nn/functional/loss.py [hunk 8]
```diff
@@ -64,13 +64,116 @@
...
+        The tensor variable storing the smooth_l1_loss of input and label.
+
+    Return type: Tensor.
+
+    Examples:
+        .. code-block:: python
+
+            # declarative mode
+            import paddle
+            import paddle.fluid as fluid
```
Comment (saxon-zh) [inline python/paddle/nn/functional/loss.py]: 需要使用2.0里面的API，去掉fluid

Comment (saxon-zh) [reply python/paddle/nn/functional/loss.py]: 这个示例里面所有调用fluid的接口都需要修改下

File: python/paddle/nn/layer/loss.py [hunk 9]
```diff
@@ -574,6 +575,87 @@ def forward(self, input, label):
             name=self._name)
 
 
+class SmoothL1Loss(fluid.dygraph.Layer):
+    """
+	:alias_main: paddle.nn.SmoothL1Loss
+	:alias: paddle.nn.SmoothL1Loss,paddle.nn.layer.SmoothL1Loss,paddle.nn.layer.loss.SmoothL1Loss
```
Comment (saxon-zh) [inline python/paddle/nn/layer/loss.py]: 去掉alias

File: python/paddle/nn/layer/loss.py [hunk 10]
```diff
@@ -574,6 +575,87 @@ def forward(self, input, label):
...
+        The tensor variable storing the smooth_l1_loss of input and label.
+
+    Return type: Tensor.
+
+    Examples:
+        .. code-block:: python
+
+            # declarative mode
+            import paddle
+            import paddle.fluid as fluid
```
Comment (saxon-zh) [inline python/paddle/nn/layer/loss.py]: 去掉fluid

Comment (yghstill) [reply python/paddle/nn/layer/loss.py]: Done, thx.

File: python/paddle/nn/layer/loss.py (hunk 9 above)
Comment (yghstill) [reply python/paddle/nn/layer/loss.py]: Done, thx.

File: python/paddle/nn/functional/loss.py (hunk 8 above)
Comment (yghstill) [reply python/paddle/nn/functional/loss.py]: Done, thx.

File: python/paddle/nn/functional/loss.py (hunk 7 above)
Comment (yghstill) [reply python/paddle/nn/functional/loss.py]: Done, thx.

File: python/paddle/nn/layer/loss.py [hunk 11]
```diff
@@ -574,6 +575,70 @@ def forward(self, input, label):
...
+        .. code-block:: python
+
+            # declarative mode
+            import paddle
+            import numpy as np
+
+            paddle.disable_static()
+            input_data = np.random.rand(3,3).astype("float32")
+            label_data = np.random.rand(3,3).astype("float32")
+            input = paddle..to_variable(input_data)
```
Comment (saxon-zh) [inline python/paddle/nn/layer/loss.py]: 多了一个.

Comment (yghstill) [reply python/paddle/nn/layer/loss.py]: Done，Thx

**Review by saxon-zh — DISMISSED** [review]
LGTM

**Review by juncaipeng — DISMISSED** [review]
LGTM

**Review by jerrywgz — DISMISSED** [review]
LGTM

**Review by Shixiaowei02 — DISMISSED** [review]

**Review by yaoxuefeng6 — DISMISSED** [review]

**Review by jerrywgz — DISMISSED** [review]

**Review by yaoxuefeng6 — DISMISSED** [review]

**Review by juncaipeng — DISMISSED** [review]
LGTM

**Review by Shixiaowei02 — DISMISSED** [review]

**Review by saxon-zh — DISMISSED** [review]
LGTM

File: python/paddle/nn/functional/loss.py [hunk 12]
```diff
@@ -64,13 +64,100 @@
...
     'softmax_with_cross_entropy',
     'square_error_cost',
     'ssd_loss',
     'teacher_student_sigmoid_loss'
 ]
 
 
+def smooth_l1_loss(input, label, reduction='mean'):
+    """
+    This operator is calculate smooth_l1_loss. Creates a criterion that uses a squared term if the absolute element-wise error falls below 1
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py]: This operator calculates smooth_l1_loss. 另外也要提一下对outliners不敏感。

File: python/paddle/nn/functional/loss.py [hunk 13]
```diff
@@ -64,13 +64,100 @@
...
+    and an L1 term otherwise. In some cases it can prevent exploding gradients. Also known as the Huber loss:
```
Comment (jzhang533) [inline python/paddle/nn/functional/loss.py]: paddle是有一个huber loss的：`paddle.fluid.layers.huber_loss`，是可以设置delta的。
可以在`paddle.fluid.layers.huber_loss`的文档里加一句话，推荐用户用`paddle.nn.functional.smooth_l1_loss`

File: python/paddle/nn/layer/loss.py [hunk 14]
```diff
@@ -711,3 +711,66 @@ def forward(self, input, other, target):
         out = paddle.nn.functional.margin_ranking_loss(
             input, other, target, self.margin, self.reduction, self.name)
         return out
+
+
+class SmoothL1Loss(fluid.dygraph.Layer):
+    """
+    This operator is calculate smooth_l1_loss. Creates a criterion that uses a squared
```
Comment (jzhang533) [inline python/paddle/nn/layer/loss.py]: This operator calculates smooth_l1_loss.

File: python/paddle/nn/layer/loss.py [hunk 15]
```diff
@@ -711,3 +711,66 @@ def forward(self, input, other, target):
...
+
+    Returns:
+        The tensor variable storing the smooth_l1_loss of input and label.
+
+    Return type: Tensor.
+
+    Examples:
+        .. code-block:: python
+
+            # declarative mode
```
Comment (jzhang533) [inline python/paddle/nn/layer/loss.py]: this is a imperative example

**Review by jzhang533 — DISMISSED** [review]
lgtm

**Review by yaoxuefeng6 — DISMISSED** [review]

File: python/paddle/nn/layer/loss.py [hunk 16]
```diff
@@ -711,3 +711,66 @@ def forward(self, input, other, target):
...
+         \\mathop{z_i}=\\left\\{\\begin{array}{rcl}
+        0.5(x_i - y_i)^2 & & {if |x_i - y_i| > 1} \\\\
+        |x_i - y_i| - 0.5 & & {otherwise}
+        \\end{array} \\right.
+
+    Parameters:
+        reduction (str, optional): Indicate how to average the loss by batch_size,
+            the candicates are ``'none'`` | ``'mean'`` | ``'sum'``.
+            If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned;
+            If :attr:`size_average` is ``'sum'``, the reduced sum loss is returned.
```
Comment (XiaoguangHu01) [inline python/paddle/nn/layer/loss.py:740]: retuction

File: python/paddle/nn/functional/loss.py [hunk 17]
```diff
@@ -64,13 +64,103 @@
...
+    if reduction not in ['sum', 'mean', 'none']:
+        raise ValueError(
+            "The value of 'reduction' in smooth_l1_loss should be 'sum', 'mean' or"
+            " 'none', but received %s, which is not allowed." % reduction)
+    if reduction == 'none':
+        return out
+    reduce_op = 'reduce_mean'
+    if reduction == 'sum':
+        reduce_op = 'reduce_sum'
+    return getattr(fluid.layers, reduce_op)(out)
```
Comment (XiaoguangHu01) [inline python/paddle/nn/functional/loss.py:161]: 使用getattr函数可能会降低性能，
这里的if else逻辑并不复杂，
建议直接使用fluid.layers.resuce_mean 和fluid.layers.reduce_sum的写法。

File: python/paddle/nn/functional/loss.py [hunk 18]
```diff
@@ -64,13 +64,103 @@
...
+    Parameters:
+        input (Tensor): Input tensor, the data type is float32. Shape is
+            (N, C), where C is number of classes, and if shape is more than 2D, this
+            is (N, C, D1, D2,..., Dk), k >= 1.
+        label (Tensor): Label tensor, the data type is float32. The shape of label
+            is the same as the shape of input.
+        reduction (str, optional): Indicate how to average the loss by batch_size,
+            the candicates are ``'none'`` | ``'mean'`` | ``'sum'``.
+            If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned;
+            If :attr:`size_average` is ``'sum'``, the reduced sum loss is returned.
```
Comment (XiaoguangHu01) [inline python/paddle/nn/functional/loss.py:105]: reduction

File: python/paddle/nn/functional/loss.py [hunk 19]
```diff
@@ -64,13 +64,103 @@
...
+
+            paddle.disable_static()
+            input_data = np.random.rand(3,3).astype("float32")
+            label_data = np.random.rand(3,3).astype("float32")
+            input = paddle.to_tensor(input_data)
+            label = paddle.to_tensor(label_data)
+            output = paddle.nn.functioanl.smooth_l1_loss(input,label)
+            print(output.numpy())
+    """
+    fluid.data_feeder.check_variable_and_dtype(input, 'input', ['float32'],
```
Comment (XiaoguangHu01) [inline python/paddle/nn/functional/loss.py:130]: 支持float32和float64

File: python/paddle/nn/functional/loss.py [hunk 20]
```diff
@@ -64,13 +64,103 @@
...
+                                               'smooth_l1_loss')
+    fluid.data_feeder.check_variable_and_dtype(label, 'label', ['float32'],
```
Comment (XiaoguangHu01) [inline python/paddle/nn/functional/loss.py:132]: 支持float32和float64

File: python/paddle/fluid/tests/unittests/test_smooth_l1_loss_v2_op.py [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+        self.check_output()
+
+    def test_check_grad_normal(self):
+        self.check_grad(['X', 'Y'], 'Out', max_relative_error=0.03)
+
+    def test_check_grad_ingore_x(self):
+        self.check_grad(
+            ['Y'],
+            'Out',
+            max_relative_error=0.03,
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/tests/unittests/test_smooth_l1_loss_v2_op.py:96]: 这里不允许放大max_relative_error的值，如果精度不满足，可以把反向检查的数据类型改为float64
https://github.com/PaddlePaddle/Paddle/wiki/OP-test-accuracy-requirements

File: python/paddle/fluid/tests/unittests/test_smooth_l1_loss.py [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+    if reduction == 'sum':
+        return np.sum(out)
+    elif reduction == 'mean':
+        return np.mean(out)
+    elif reduction == 'none':
+        return out
+
+
+class SmoothL1Loss(unittest.TestCase):
+    def test_smooth_l1_loss_mean(self):
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/tests/unittests/test_smooth_l1_loss.py:43]: 可以再setup函数设置下numpy的seed，避免随机性引起的单测随机挂
